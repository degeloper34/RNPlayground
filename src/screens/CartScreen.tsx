import {FontAwesome} from "@expo/vector-icons";
import {useContext} from "react";
import {FlatList, Image, Pressable, StyleSheet, View} from "react-native";
import {RootStackScreenProps} from "../../types";
import {CustomText, EmptyState} from "../components/atoms";
import Colors from "../constants/Colors";
import {MainContext} from "../context/mainContext";

export default function CartScreen({navigation}: RootStackScreenProps<"Cart">) {
  const context = useContext(MainContext);
  const showEmptyState = Object.keys(context.cart).length === 0;

  const renderItem = ({item}: {item: string}) => {
    return (
      <View style={styles.renderItemView}>
        <Image
          source={{uri: context.cart[item].image}}
          style={styles.imageStyle}
          resizeMode={"contain"}
        />
        <View style={styles.renderItemInnerView}>
          <CustomText
            text={context.cart[item].title}
            type={"medium"}
            numberOfLines={1}
          />
          <View style={styles.marginTop} />
          <CustomText
            text={"Unit Price: $" + context.cart[item].price}
            type={"medium"}
          />
          <View style={styles.marginTop} />
          <CustomText
            text={"Quantity: " + context.cart[item].quantity}
            type={"medium"}
          />
          <View style={styles.marginTop} />
          <CustomText
            text={
              "Total Price: $" +
              context.cart[item].quantity * context.cart[item].price
            }
            type={"medium"}
          />
        </View>
        <Pressable
          onPress={() =>
            context.updateCart(
              context.cart[item],
              context.cart[item].quantity,
              "remove"
            )
          }
        >
          <FontAwesome name="remove" size={24} color="black" />
        </Pressable>
      </View>
    );
  };

  return (
    <View style={styles.container}>
      {showEmptyState ? (
        <EmptyState emptyStateText={"There is no product in your cart."} />
      ) : (
        <FlatList
          data={Object.keys(context.cart)}
          renderItem={renderItem}
          style={styles.flatListStyle}
          keyExtractor={(_, index) => String(index)}
        />
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    backgroundColor: Colors.white,
  },
  renderItemView: {
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: Colors.mainGrey,
    padding: 6,
    borderRadius: 8,
    marginBottom: 12,
  },
  imageStyle: {
    width: 70,
    height: 70,
  },
  renderItemInnerView: {
    paddingHorizontal: 12,
    flex: 1,
  },
  marginTop: {
    marginTop: 6,
  },
  flatListStyle: {
    padding: 12,
  },
});
