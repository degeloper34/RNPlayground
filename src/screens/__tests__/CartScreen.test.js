import renderer from "react-test-renderer";
import React from "react";
import CartScreen from "../CartScreen";

jest.mock("@expo/vector-icons");

describe("CartScreen", () => {
  it("should render without issues", () => {
    const screen = renderer.create(<CartScreen />).toJSON();

    expect(screen).toMatchSnapshot();
  });
});
