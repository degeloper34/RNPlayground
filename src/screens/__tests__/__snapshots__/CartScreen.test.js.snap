// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartScreen should render without issues 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          undefined,
          Object {
            "color": "#21325E",
            "fontFamily": "sora-medium",
            "fontSize": 14,
            "textDecorationLine": "none",
          },
        ]
      }
    >
      There is no product in your cart.
    </Text>
  </View>
</View>
`;
